src/os/bluestore/BlueStore.cc

```diff
@@ -6209,38 +6337,33 @@ int BlueStore::_do_write(
 	   << " comp_blob_size 0x" << std::hex << wctx.comp_blob_size
 	   << std::dec << dendl;
 
-  bufferlist::iterator p = bl.begin();
-  if (offset / min_alloc_size == (end - 1) / min_alloc_size &&
-      (length != min_alloc_size)) {
-    // we fall within the same block
-    _do_write_small(txc, c, o, offset, length, p, &wctx);
-  } else {
-    uint64_t head_offset, head_length;
-    uint64_t middle_offset, middle_length;
-    uint64_t tail_offset, tail_length;
-
-    head_offset = offset;
-    head_length = P2NPHASE(offset, min_alloc_size);
+  uint64_t gc_start_offset = offset, gc_end_offset = end;
 
-    tail_offset = P2ALIGN(end, min_alloc_size);
-    tail_length = P2PHASE(end, min_alloc_size);
+  if (blobs_need_garbage_collection(o, offset, end, &wctx.le_depth,
+                            &gc_start_offset, &gc_end_offset) == true) {
+    // we need garbage collection of blobs.
+    wctx.gc_write = true;
 
-    middle_offset = head_offset + head_length;
-    middle_length = length - head_length - tail_length;
-
-    if (head_length) {
-      _do_write_small(txc, c, o, head_offset, head_length, p, &wctx);
+    if (offset > gc_start_offset) {
+      bufferlist head_bl;
+      size_t read_len = offset - gc_start_offset;
+      int r = _do_read(c.get(), o, gc_start_offset, read_len, head_bl, 0);
+      assert(r >= 0 && r <= (int)read_len); 
+      _do_write_data(txc, c, o, gc_start_offset, read_len, head_bl, &wctx);
```

Comment:
By doing separate writes for head, tail and body you considerably reduce the benefit that collector is  expected to bring. E.g. one can have following writes:
1) 0~128K, 128K~128K
2) 64K~128K
3) 64K~64K
Resulting lextents' map will be (IMHO):
0~64, 64~64, 128~128
Not a grate achievement for a costly GC operation.